doc/b2pf.3
.TH B2PF 3 "09 April 2025" "B2PF 0.11"
.SH NAME
B2PF - Unicode base code to presentation forms converter
.sp
.B #include <b2pf.h>
.sp
B2PF is a library of C functions for converting "base" characters in Unicode
strings into "presentation form" characters for display or printing. The
library was inspired by the need to do this for characters in Arabic scripts,
but as it is configurable, it can be used (with an appropriate configuration)
in any situation where this kind of transformation is required. A basic Arabic
configuration is supplied with the distribution.
.P
B2PF supports UTF-8, UTF-16, and UTF-32 strings, and can handle input and
output in reading order or reverse reading order.
.
.
.SH "B2PF FUNCTIONS"
.rs
.sp
.nf
.B int b2pf_context_create(const char *\fIrules_name\fP,
.B "  const char *\fIrules_dir_list\fP, uint32_t \fIoptions\fP,"
.B "  b2pf_context **\fIcontextptr\fP,"
.B "  void *(*\fIprivate_malloc\fP)(size_t, void *),"
.B "  void *(*\fIprivate_free\fP)(void *, void *), void *\fImemory_data\fP,"
.B "  unsigned int *\fIlinenumberptr\fP);"
.sp
.B int b2pf_context_add_file(b2pf_context *\fIcontext\fP,
.B "  const char *\fIrules_name\fP, const char *\fIrules_dir_list\fP,"
.B "  uint32_t \fIoptions\fP, unsigned int *\fIlinenumberptr\fP);"
.sp
.B int b2pf_context_add_line(b2pf_context *\fIcontext\fP, const char *\fIrule_line\fP);
.sp
.B int b2pf_context_set_callback(b2pf_context *\fIcontext\fP, uint32_t \fIoptions\fP,
.B "  int(*\fIcallback\fP)(uint32_t, void *), void *\fIdata\fP);"
.sp
.B void b2pf_context_free(b2pf_context *\fIcontext\fP);
.sp
.B int b2pf_format_string(b2pf_context *\fIcontext\fP, void *\fIinput_string\fP,
.B "  size_t \fIinput_size\fP, void *\fIoutput_string\fP, size_t \fIoutput_size\fP,"
.B "  size_t *\fIoutput_used\fP, uint32_t \fIoptions\fP, size_t *\fIerror_offset\fP);"
.sp
.B int b2pf_get_error_message(int \fIerrorcode\fP, void *\fImessage_buffer\fP,
.B "  size_t \fIbuffer_size\fP, size_t *\fIbuffer_used\fP, uint32_t \fIoptions\fP);"
.sp
.B int b2pf_get_check_message(b2pf_context *\fIcontext\fP, void *\fImessage_buffer\fP,
.B "  size_t \fIbuffer_size\fP, size_t *\fIbuffer_used\fP, uint32_t \fIoptions\fP);"
.fi
.
.
.SH "B2PF OVERVIEW"
.rs
.sp
B2PF uses a "context" to store information about the Unicode characters of
interest and the rules by which words are to be transformed. The first thing a
calling program must do is to create a context and populate it. Then the main
processing function, \fBb2pf_format_string()\fP, can be used to convert
character strings to their presentation forms.
.P
Processing happens in reading order, that is, the order in which characters are
normally read. For a Roman script such as English, this is left-to-right,
whereas for Arabic and similar scripts it is right-to-left. By default, the
input characters are assumed to be in reading order, but there are options that
specify two different forms of reverse reading order. In these cases, the
string is reversed before processing. It is also converted to UTF-32 if
necessary, because this is the form in which it is processed. All input is
checked for UTF validity.
.P
The input string is processed word by word. Characters that cannot be part of a
word (as specified in the
.\" HTML <a href="#rules">
.\" </a>
context)
.\"
are copied unchanged to the output. Each word that is identified is processed
independently. When the end of the string is reached, the output string is
reversed if one of the options for non-reading-order output is set, and it is
then converted to UTF-16 or UTF-8 if necessary.
.
.
.SS "Memory management"
.rs
.sp
B2PF uses heap memory for storing contexts, and it may also use it if an input
string is too long for the internal working buffers that are on the stack.
There is provision for the use of a custom memory allocator; if one is not
supplied, the system \fBmalloc()\fP and \fBfree()\fP functions are used.
.
.
.SH "CREATING AND DESTROYING A CONTEXT"
.rs
.sp
.nf
.B int b2pf_context_create(const char *\fIrules_name\fP,
.B "  const char *\fIrules_dir_list\fP, uint32_t \fIoptions\fP,"
.B "  b2pf_context **\fIcontextptr\fP,"
.B "  void *(*\fIprivate_malloc\fP)(size_t, void *),"
.B "  void *(*\fIprivate_free\fP)(void *, void *), void *\fImemory_data\fP,"
.B "  unsigned int *\fIlinenumberptr\fP);"
.sp
.B void b2pf_context_free(b2pf_context *\fIcontext\fP);
.fi
.sp
The \fBb2pf_context_create()\fP function creates a new context. If its first
argument is NULL or an empty string, an empty context is created. Information
can be added by calling \fBb2pf_context_add_file()\fP or
\fBb2pf_context_add_line()\fP. If not NULL or an empty string, the first
argument must be the name of a rules file, whose format is defined in the
.\" HTML <a href="#rules">
.\" </a>
section entitled "Creating Rules"
.\"
below. If the second argument is not NULL, it must be a colon-separated list of
directories that are to be searched for the rules file, before the default
rules directory is searched. This argument is ignored if the first argument is
NULL or an empty string.
.P
The third argument contains option bit settings. No options are currently
defined; this is a placeholder in the API for possible future use.
.P
The fourth argument is a pointer to a variable of type \fBb2pf_context *\fP,
through which the address of the newly-created context is returned.
.P
The next three arguments are concerned with custom memory management, and
may all be NULL if the system \fBmalloc()\fP and \fBfree()\fP functions are to
be used. Otherwise, the first of these three must point to a memory allocator;
its arguments are the size required and an opaque value that can be used to
distinguish the source of the call. Similarly, the second argument points to a
de-allocator, again with an opaque identifier. When either of these functions
are called, the value given for the \fImemory_data\fP argument is passed as the
opaque value.
.P
The final argument for \fBb2pf_context_create()\fP is a pointer to an unsigned
integer, into which a line number is placed in the event of an error while
processing a rules file.
.P
If \fBb2pf_context_create()\fP is successful, it returns B2PF_SUCCESS and
places a pointer to the new context in the variable pointed to by
\fIcontextptr\fP. Otherwise, the value returned is an error code. The
\fBb2pf_get_error_message()\fP function can be used to obtain a textual error
message. For more details of errors, see the
.\" HTML <a href="#errors">
.\" </a>
section entitled "Handling errors"
.\"
below.
.P
When a context is no longer needed, its memory should be released by calling
\fBb2pf_context_free()\fP, whose only argument is a pointer to the context.
.
.
.SH "EXTENDING A CONTEXT"
.rs
.sp
.nf
.B int b2pf_context_add_file(b2pf_context *\fIcontext\fP,
.B "  const char *\fIrules_name\fP, const char *\fIrules_dir_list\fP,"
.B "  uint32_t \fIoptions\fP, unsigned int *\fIlinenumberptr\fP);"
.sp
.B int b2pf_context_add_line(b2pf_context *\fIcontext\fP, const char *\fIrule_line\fP);
.fi
.sp
Information can be added to a context by calling \fBpcre2_context_add_file()\fP
or \fBb2pf_context_add_line()\fP. The first of these reads a rules file and
adds its information to the context that is its first argument. The other
arguments, \fIrules_name\fP, \fIrules_dir_list\fP, \fIoptions\fP, and
\fIlinenumberptr\fP, are the same as for \fBb2pf_context_create()\fP, as is the
returned value.
.P
The \fBb2pf_context_add_line()\fP function's first argument is a pointer to a
context, and the second is a string in
UTF-8 code that is treated in the same way as a line in a rules file; see the
.\" HTML <a href="#rules">
.\" </a>
section entitled "Creating Rules"
.\"
below for details. By using this function and starting from an empty context,
all the rules information in a context can be provided from sources other than
files. If the rules line is accepted, the function returns B2PF_SUCCESS. If
not, the result is an error code.
.
.
.SH "SETTING UP A CALLLBACK"
.rs
.sp
.nf
.B int b2pf_context_set_callback(b2pf_context *\fIcontext\fP, uint32_t \fIoptions\fP,
.B "  int(*\fIcallback\fP)(uint32_t, void *), void *\fIdata\fP);"
.fi
.sp
B2PF rules may specify that certain pairs of characters are to be replaced by
\fIligatures\fP (for example, the single character "fi" replacing "f" followed
by "i"). However, some fonts may not contain all the specified ligatures. There
is a \fIcallback\fP facility to allow B2PF to ask the calling application
whether or not a given ligature is acceptable. This is enabled by calling
\fBb2pf_set_callback()\fP, which has four arguments.
.P
The first argument is a context pointer. The second is a set of options bits.
The only option currently defined is B2PF_CALLBACK_LIGATURE, which requests
a callback for any potential ligature. All other bits must be zero; they allow
for possible future callback extensions. The third argument is a pointer to a
callback function, and the final argument is opaque data to be passed to that
function.
.P
When the B2PF_CALLBACK_LIGATURE option is set, whenever B2PF is about to
replace characters in the input by a ligature character, it calls the callback
function with the ligature's Unicode codepoint as the first argument, and the
opaque data that was passed to \fBb2pf_set_callback()\fP as the second. The
function must return zero if the ligature is unacceptable or non-zero
otherwise.
.P
Callbacks can be disabled by calling \fBb2pf_set_callback()\fP with the
\fIoptions\fP argument zero; in this case the function argument may be NULL.
The result of calling this function is zero if all went well or else an error
code.
.
.
.SH "FORMATTING A STRING"
.rs
.sp
.nf
.B int b2pf_format_string(b2pf_context *\fIcontext\fP, void *\fIinput_string\fP,
.B "  size_t \fIinput_size\fP, void *\fIoutput_string\fP, size_t \fIoutput_size\fP,"
.B "  size_t *\fIoutput_used\fP, uint32_t \fIoptions\fP, size_t *\fIerror_offset\fP);"
.fi
.sp
This is the function that provides the functionality of B2PF. Its first
argument is a context. The next two are a pointer to the input string and its
size. These are followed by a pointer to a buffer to receive the output, the
size of this buffer, and a pointer to a variable into which the size that is
actually used is placed. All sizes are in code units. The pointers are defined
as \fBvoid *\fP because they may point to UTF-8, UTF-16, or UTF-32 strings, as
specified by what is set in the \fBoptions\fP argument (see below). The final
argument is a pointer to a variable into which, in the event of an error, a
code unit offset into the input is placed.
.P
The function returns B2PF_SUCCESS or an error code. Because the information in
a context can be added in any order, a check on the consistency of a context
cannot be done until \fBb2pf_format_string()\fP is called. If an inconsistency
is found, the function still does its processing, but it returns
B2PF_ERROR_CONTEXTCHECK instead of B2PF_SUCCESS. You can ignore this error, but
if you want more details, see the \fBb2pf_get_check_message()\fP function in
the
.\" HTML <a href="#errors">
.\" </a>
section entitled "Handling errors"
.\"
below. An example of an inconsistency in a context is a ligature definition
where the type of one of the specified characters has not been defined.
.
.
.SS "Formatting options"
.rs
.sp
The following bit values may be set in the \fBoptions\fP argument:
.sp
  B2PF_UTF_16
  B2PF_UTF_32
.sp
At most, only one of these options may be set. They specify, respectively, that
the input is coded as UTF-16 or UTF-32 code units, and the output is to be in
the same encoding. If neither is set, UTF-8 is assumed.
.sp
  B2PF_INPUT_BACKCHARS
  B2PF_INPUT_BACKCODES
.sp
At most, only one of these options may be set. They specify that the input is
in reverse reading order, but they differ in their effect when the input
contains combining characters such as diacriticals. In the reading order in
which B2PF does its processing, combining characters must follow the character
to which they apply. If B2PF_INPUT_BACKCHARS is set, combining characters are
assumed to follow their principal character in the input, and therefore the
string is reversed by treating each character and its following combiners as a
unit. In contrast, if B2PF_INPUT_BACKCODES is set, an input string is reversed
code unit by code unit. This is appropriate if the input has combining
characters preceding the characters with which they combine.
.sp
  B2PF_OUTPUT_BACKCHARS
  B2PF_OUTPUT_BACKCODES
.sp
At most, only one of these options may be set. They affect the output string in
the same ways as the input options above.
.
.
.\" HTML <a name="errors"></a>
.SH "HANDLING ERRORS"
.rs
.sp
.nf
.B int b2pf_get_error_message(int \fIerrorcode\fP, void *\fImessage_buffer\fP,
.B "  size_t \fIbuffer_size\fP, size_t *\fIbuffer_used\fP, uint32_t \fIoptions\fP);
.sp
.B int b2pf_get_check_message(b2pf_context *\fIcontext\fP, void *\fImessage_buffer\fP,
.B "  size_t \fIbuffer_size\fP, size_t *\fIbuffer_used\fP, uint32_t \fIoptions\fP);
.fi
.sp
Error codes may be positive or negative. Negative error codes are used when a
UTF input string fails the validity check. Each error code has a name that is
defined in \fBb2pf.h\fP. A textual error message can be obtained by calling
\fBb2pf_get_error_message()\fP. Its first argument is an error code, the second
is a pointer to a buffer that is to receive the message, and the third is the
buffer size in code units. The fourth argument is used to return the number of
code units that are used. The final argument may contain one of the values
B2PF_UTF_16 or B2PF_UTF_32, requesting that the output be in UTF-16 or UTF-32
encoding, respectively. The default is to return UTF-8, but in practice all the
error texts contain only ASCII characters, so the only difference the options
make is to specify the bit-size of the code units. The function returns an
error code if the error number or options bits are invalid or the output buffer
is too small.
.P
The error B2PF_ERROR_CONTEXTCHECK means that there is some inconsistency in the
context, though processing otherwise succeeded. For example, a ligature has
been defined for characters whose type has not been declared, meaning they can
never be part of a word. The \fBb2pf_get_check_message()\fP function can be
used to get a message that gives details of the inconsistency. Its arguments
are the same as \fBb2pf_get_error_message()\fP, except that the first argument
is a pointer to the context instead of an error code.
.
.
.\" HTML <a name="rules"></a>
.SH "CREATING RULES"
.rs
.sp
Information in a context is of two types, character definitions and processing
rules. The lines of text that specify this information can be read from a rules
file that is named in a call to \fBb2pf_context_create()\fP or
\fBb2pf_context_add_file()\fP, or added line by line by calls to
\fBb2pf_context_add_line()\fP. In all cases the lines must be encoded in UTF-8.
.P
Each type of line may be repeated as many times as necessary, and the
information may be specified in any order. However, when
\fBb2pf_format_string()\fP is called, the processing rules are obeyed in the
order in which they are defined. There is also a check on the consistency of
the context at this time.
.P
Empty lines and lines that start with # are treated as comments and ignored.
All other lines must start with one of six capital letters, followed by at
least one space. In the description below, wherever a character definition is
required, either a literal UTF-8 character or a Unicode code point given as U+
followed by a hexadecimal number may be given.
.
.
.SS "Define combining characters"
.rs
.sp
A line beginning with C defines characters that combine with their preceding
characters to form parts of words (for example, diacriticals). The data is a
list of individual characters or ranges separated by a hyphen. White space is
ignored, but not allowed within a range. If # appears, it and any following
text is ignored. For example:
.sp
  C U+064B-U+0655 U+0670  # Arabic diacritics
.sp
.
.
.SS "Define miscellaneous word characters"
.rs
.sp
A line beginning with M defines characters that can be part of words, but are
not combining characters and do not have multiple presentation forms. Its
format is the same as the C rule, that is, M is followed by list of individual
characters and/or character ranges.
.
.
.SS "Define a character with presentation forms"
.rs
.sp
A line beginning with P defines a character that has multiple presentation
forms. P must be followed (left to right) by five characters; there may also be
an optional comment. The five characters are, in order, the base character and
its isolated, initial, medial, and final forms. If any of these do not exist, a
single hyphen is used as a place marker. Here are two Arabic examples:
.sp
  P U+0627   U+FE8D   -        -        U+FE8E  # ALEF
  P U+0628   U+FE8F   U+FE91   U+FE92   U+FE90  # BEH
.sp
ALEF (U+0627) has only an isolated and a final form, whereas BEH (U+0628) has
all four presentation forms.
.
.
.SS "Define an input ligature"
.rs
.sp
A line beginning with L defines a ligature, that is, a pair of adjacent
characters that are to be replaced by a single combined code point. This
happens as a word is being identified, before it is processed using the rules.
There is also a facility for replacing pairs of characters after processing
(see below).
.P
The data for an L line is the three participating characters, optionally
followed by a comment. For example:
.sp
  L f      i      U+FB01   # Latin fi ligature
  L U+0644 U+0627 U+FEFB   # Arabic LAM with ALEF
.sp
Reading left to right, the first two characters after L define the ligature, in
reading order, and the third is the replacement character. In the Arabic
example, LAM (U+0644) would appear to the right of ALEF (U+0627) in print that
is read right-to-left. The replacement character need not appear in a C, M, or
P line; if it does not, it is treated as a "miscellaneous" character.
.P
If the replacement character has multiple presentation forms, it should also
appear in a P line (see above). A ligature character itself may be the first
character in another ligature, so it is possible to define a combined ligature
of three (or more) characters.
.P
Ligatures are recognized only for two combining characters (defined by C
lines), or for two non-combining characters (defined by M or P lines). A
ligature definition that lists one character of each type or contains an
undefined character has no effect and causes the B2PF_ERROR_CONTEXTCHECK
error code to be returned by \fBb2pf_format_string()\fP.
.P
When a non-combining character is followed by a combining character (for
example, a letter followed by a diacritic), the next non-combining character is
considered when checking for ligatures. In the Arabic example above, the
LAM/ALEF ligature is recognized even if there is a diacritic between LAM and
ALEF. In effect, the sequence LAM, diacritic(s), ALEF is processed as if it
were LAM, ALEF, diacritic(s).
.
.
.SS "Define an output ligature"
.rs
.sp
A line beginning with A (for "afterwards") defines a ligature that is
recognized after a word has been processed. Such ligatures usually involve the
presentation forms of characters, which is why they cannot be recognized
earlier. The format of an A line is the same as an L line. For example:
.sp
  A U+FE91 U+FEAE U+FC6A  # BEH initial + REH final
  A U+FE92 U+FEAE U+FC6A  # BEH medial + REH final
.sp
Output ligatures are supported only for non-combining characters, but, as for L
ligatures, combining characters may intervene. If either of the first two
characters is not defined in a C, M, or P line, it is treated as a
non-combining character.
.
.
.SS "Specify a processing rule"
.rs
.sp
The actual rules for processing are specified by R lines. Words are scanned
character by character in reading order, and for each character the rules are
tried in order until one matches. No further rules are then considered for the
current character. The format of an R line is as follows:
.sp
  R [<\fIoptions\fP>] <\fIpattern\fP> -> <\fIreplacement\fP>
.sp
White space is entirely ignored in R lines. Options (which are optional) must
appear in square brackets immediately after the R. At present no options are
defined and empty square bracket can be omitted; this syntax is a placeholder
for possible future development.
.P
The pattern is NOT a regular expression, though some of its syntax looks
similar. It is divided into three parts by parentheses, as in this example:
.sp
  R \eN(\ei)\ep -> \ei
.sp
There may be only one set of parentheses in the pattern. Inside the parentheses
is a definition of one or more characters that are to be replaced. The first
part, before the opening parenthesis, defines a pre-assertion, that is,
something that must match before the replaceable characters, and similarly the
final part is a post-assertion. In the example above the pre-assertion is \eN
and the post-assertion is \ep.
.P
In each of these parts, characters stand for themselves except for the
following:
.sp
  #    Marks the end of the rule and introduces comment
  ->   Marks the end of the pattern
  ^    Asserts start of word
  $    Asserts end of word
  .    Matches any character
  \e    Introduces an escape sequence (see below)
.sp
The characters ^ and $ may appear only at the start or end of the pattern,
respectively. The dot character matches any character, together with any
following combining characters.
.P
A backslash followed by a non-alphanumeric character encodes the second
character as a literal (though including non-alphanumeric characters in words
seems an unlikely scenario). A backslash followed by a letter introduces one of
the following escape sequences that match characters of certain types:
.sp
  \ef  Character has a final form
  \ei  Character has an initial form
  \em  Character has a medial form
  \en  Character can join to the next character
  \eN  Character cannot join to the next character
  \ep  Character can join to the previous character
  \eP  Character cannot join to the previous character
  \es  Character has an isolated (solitary) form
  \eU  Specify literal character by Unicode code point
.sp
Whenever a character is matched, any following combining characters are
included in the match. \eU must be followed by + and a hexadecimal code point
value. If a terminator is needed because the next character happens to be a
hexadecimal digit, a space can be used because white space is ignored in rules.
.P
In the replacement text, # ends the text and introduces comment. A full stop
and various escape sequences are recognized as special. Except for \eU+, which
defines a literal character by Unicode code point, whenever any one of these is
encountered, B2PF looks at the next character in the word that matched a dot or
one of the character type escape sequences within the parenthesized part of the
pattern, and inserts one of its forms, as follows:
.sp
  .   The base form
  \ef  The final form
  \ei  The initial form
  \em  The medial form
  \es  The isolated (solitary) form
.sp
If the character does not have the requested presentation form, or too few
non-literal characters were matched within parentheses, an error occurs. The
escape sequences \en, \eN, \ep, and \eP are invalid in replacement strings.
.P
Consider the example quoted above:
.sp
  R \eN(\ei)\ep -> \ei
.sp
This example specifies that a character possessing an initial form that is
preceded by a character that does not join to its successor, but is followed by
a character that joins to its predecessor, should be replaced by its initial
form. Note that this requires the modified character to have a predecessor. In
practice, the above rule would normally be accompanied by another one, which
covers the case when the character in question is at the start of a word:
.sp
  R ^(\ei)\ep -> \ei
.sp
Other examples can be seen in the Arabic rules file.
.
.
.SH "SUPPLIED RULES FILES"
.rs
.sp
Some rules files are distributed in the "rules" directory, which is installed
as <something>/share/b2pf/rules, where <something> is often /usr/local. The
following files are currently included:
.P
\fBArabic\fP: This file contains definitions of base letters and diacritics in
the Arabic script, and rules for joining letters. It also contains definitions
of the basic LAM+ALEF ligatures that are usually found in Arabic fonts.
.P
\fBArabic-LigExtra\fP: This file contains definitions of a number of other
ligatures that are defined by Unicode, but which are not always present in
Arabic fonts.
.
.
.SH "SEE ALSO"
.rs
.sp
\fBb2pf-config\fP(1), \fBb2pftest\fP(1)
.
.
.SH AUTHOR
.rs
.sp
.nf
Philip Hazel
Cambridge, England.
.fi
.
.
.SH REVISION
.rs
.sp
.nf
Last updated: 09 April 2025
Copyright (c) 2025 Philip Hazel
.fi


NEWS
News about B2PF releases
------------------------

Version 0.11 09-April-2025
--------------------------

Added the callback facility for checking ligatures.


Version 0.10 23-September-2020
------------------------------

Added more apparatus for handling ligatures. See ChangeLog for details.


Version 0.00 06-August-2020
---------------------------

This is the first release.

****


src/b2pf.c
/*************************************************
*       Base Unicode to Presentation Forms       *
*************************************************/

/* This file contains the main processing function for the B2PF library.

                 Copyright (c) 2020 Philip Hazel

-----------------------------------------------------------------------------
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice,
      this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

    * The names of any contributors to this project may not be used to
      endorse or promote products derived from this software without
      specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
-----------------------------------------------------------------------------
*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "b2pf_internal.h"

#define UTF8   1        /* Number of bytes per code unit */
#define UTF16  2
#define UTF32  4

#define KNOWN_OPTIONS (B2PF_UTF_16|B2PF_UTF_32|B2PF_INPUT_BACKCHARS| \
  B2PF_INPUT_BACKCODES|B2PF_OUTPUT_BACKCHARS|B2PF_OUTPUT_BACKCODES)


/*************************************************
*             Invert a 32-bit vector             *
*************************************************/

/* Re-ordering can be by characters (that is, a base + optional combiners), or
by codes, which is by individual code units.

Arguments:
  p        points to the vector
  size     number of elements
  bychars  if TRUE, re-order characters with combiners
  context  the current context

Returns:   nothing
*/

static void
invert(uint32_t *p, size_t size, BOOL bychars, b2pf_context *context)
{
size_t i = 0;
size_t j = size - 1;

/* Inverting by code points is straightforward and we do that first. */

while (j > i)
  {
  size_t temp = p[i];
  p[i++] = p[j];
  p[j--] = temp;
  }

if (!bychars) return;

/* Inverting by character requires checking for combining characters and moving
them to after the following non-combiner. */

for (i = 0; i < size; i++)
  {
  tree_node *t = PRIV(tree_search)(context->chartreebase, p[i]);

  if (t == NULL || t->type != CT_COMB) continue;

  for(j = i + 1; j < size; j++)
    {
    t = PRIV(tree_search)(context->chartreebase, p[j]);
    if (t == NULL || t->type != CT_COMB) break;
    }

  /* If j is at the end, there is no base character. Just ignore anomalous
  case. Otherwise, pick up the base character, shift up all the combiners, and
  put the base at the start. */

  if (j < size)
    {
    size_t k;
    uint32_t c = p[j];
    for (k = j; k > i; k--) p[k] = p[k-1];
    p[i] = c;
    }

  i = j + 1;  /* Next character to consider. */
  }
}



/*************************************************
*        Main action: format a UTF string        *
*************************************************/

B2PF_EXP_DEFN int
b2pf_format_string(b2pf_context *context, void *input_string, size_t input_size,
  void *output_string, size_t output_size, size_t *output_used,
  uint32_t options, size_t *error_offset)
{
int yield = B2PF_SUCCESS;
int mode;
size_t insize, outsize, outused;
uint32_t *inbuffer = NULL;
uint32_t *outbuffer = NULL;
uint32_t stack_inbuffer[STACK_BUFFSIZE] B2PF_KEEP_UNINITIALIZED;
uint32_t stack_outbuffer[STACK_BUFFSIZE]B2PF_KEEP_UNINITIALIZED;

/* Plausibility checks */

if (context == NULL || input_string == NULL || output_string == NULL ||
    output_used == NULL || error_offset == NULL) return B2PF_ERROR_NULL;

if ((options & ~KNOWN_OPTIONS) != 0 ||
    (options & (B2PF_UTF_16|B2PF_UTF_32)) == (B2PF_UTF_16|B2PF_UTF_32) ||
    (options & (B2PF_INPUT_BACKCODES|B2PF_INPUT_BACKCHARS)) ==
               (B2PF_INPUT_BACKCODES|B2PF_INPUT_BACKCHARS) ||
    (options & (B2PF_OUTPUT_BACKCODES|B2PF_OUTPUT_BACKCHARS)) ==
               (B2PF_OUTPUT_BACKCODES|B2PF_OUTPUT_BACKCHARS))
  return B2PF_ERROR_BADOPTIONS;

/* Set up */

*error_offset = 0;
mode = ((options & B2PF_UTF_32) != 0)? UTF32 :
       ((options & B2PF_UTF_16) != 0)? UTF16 : UTF8;

/* For each code unit width, validate the UTF input/ */

if (mode == UTF8)
  yield = PRIV(valid_utf8)((uint8_t *)input_string, input_size,
    error_offset);
else if (mode == UTF16)
  yield = PRIV(valid_utf16)((uint16_t *)input_string, input_size,
    error_offset);
else   /* UTF-32 */
  yield = PRIV(valid_utf32)((uint32_t *)input_string, input_size,
    error_offset);
if (yield != B2PF_SUCCESS) goto EXIT;

/* Sort out the buffers. For UTF-8 and UTF-16, input must be converted to
UTF-32 in a local buffer. We also copy UTF-32 input so that we can shuffle the
characters if necessary when handling ligatures. If there are more input code
units than available elements in the on-stack input buffer, get memory. This
means we are safe, even when every input code unit codes for one character.
It's overkill for other cases, but does no harm (other than using more than
minimal resources). In practice, STACK_BUFFSIZE should be large enough to cater
for the majority of cases. */

if (input_size > STACK_BUFFSIZE)
  {
  inbuffer = context->malloc(input_size * sizeof(uint32_t),
    context->memory_data);
  if (inbuffer == NULL)
    {
    yield = B2PF_ERROR_MEMORY;
    goto EXIT;
    }
  }
else
  {
  inbuffer = stack_inbuffer;
  }

/* A 32-bit local output buffer is also required, except for UTF-32, when the
argument buffer can be used. For UTF-8 and UTF-16, if we got memory for an
input buffer, also get an output buffer of the same size. There's no need for
UTF-32; a small output buffer will be discovered during processing instead of
afterwards, but that's OK. */

if (mode == UTF32)
  {
  outbuffer = (uint32_t *)output_string;
  outsize = output_size;
  }
else if (inbuffer != stack_inbuffer)
  {
  outbuffer = context->malloc(input_size * sizeof(uint32_t),
    context->memory_data);
  if (outbuffer == NULL)
    {
    yield = B2PF_ERROR_MEMORY;
    goto EXIT;
    }
  outsize = input_size;
  }
else
  {
  outbuffer = stack_outbuffer;
  outsize = STACK_BUFFSIZE;
  }

/* Decode the input string into 32-bit code points. */

insize = 0;

if (mode == UTF8)
  {
  uint8_t *p8 = (uint8_t *)input_string;
  uint8_t *p8end = p8 + input_size;
  while (p8 < p8end)
    {
    uint32_t c;
    GETUTF8INC(c, p8);
    inbuffer[insize++] = c;
    }
  }

else if (mode == UTF16)
  {
  size_t i;
  uint16_t *p16 = (uint16_t *)input_string;

  for (i = 0; i < input_size; i++)
    {
    uint32_t c = *p16++;
    if ((c & 0xfc00u) == 0xd800u)
      {
      c = (((c & 0x3ffu) << 10) | (*p16++ & 0x3ffu)) + 0x10000u;
      i++;
      }
    inbuffer[insize++] = c;
    }
  }

else  /* UTF-32 */
  {
  memcpy(inbuffer, input_string, sizeof(uint32_t)*input_size);
  insize = input_size;
  }

/* If the input is backwards, invert it, then process it, and invert the output
if necessary. */

if ((options & (B2PF_INPUT_BACKCHARS|B2PF_INPUT_BACKCODES)) != 0)
  invert(inbuffer, insize, (options & B2PF_INPUT_BACKCHARS) != 0, context);

yield = PRIV(format_string)(inbuffer, insize, outbuffer, outsize, &outused,
  context, error_offset);
if (yield != B2PF_SUCCESS && yield != B2PF_ERROR_CONTEXTCHECK) goto EXIT;

if ((options & (B2PF_OUTPUT_BACKCHARS|B2PF_OUTPUT_BACKCODES)) != 0)
  invert(outbuffer, outused, (options & B2PF_OUTPUT_BACKCHARS) != 0, context);

/* Copy the output back to UTF-32, UTF-16 or UTF-8. In the UTF-32 case, the
copy is needed only if we used an internal buffer. */

if (mode == UTF32)
  {
  if (outbuffer != output_string)
    {
    if (outused > output_size) goto OVERFLOW;
    memcpy(output_string, outbuffer, sizeof(uint32_t)*outused);
    }
  *output_used = outused;
  }

else if (mode == UTF16)
  {
  size_t i;
  size_t used = 0;
  uint16_t *p16 = (uint16_t *)output_string;
  for (i = 0; i < outused; i++)
    {
    uint32_t c = outbuffer[i];
    if (c <= 0xffff)
      {
      if (used++ >= output_size) goto OVERFLOW;
      *p16++ = (uint16_t)c;
      }
    else
      {
      if (used >= output_size - 1) goto OVERFLOW;
      used += 2;
      c -= 0x10000;
      *p16++ = 0xd800 | (c >> 10);
      *p16++ = 0xdc00 | (c & 0x3ff);
      }
    }
  *output_used = used;
  }

else  /* UTF-8 */
  {
  size_t i;
  size_t used = 0;
  uint8_t *p8 = (uint8_t *)output_string;

  for (i = 0; i < outused; i++)
    {
    int j, k;
    uint8_t *pt;
    uint32_t c = outbuffer[i];
    for (j = 0; j < PRIV(utf8_table1_size); j++)
      if ((int)c <= PRIV(utf8_table1)[j]) break;
    if (used >= output_size - j) goto OVERFLOW;
    used += j + 1;
    pt = p8 += j;
    for (k = j; k > 0; k--)
     {
     *pt-- = 0x80 | (c & 0x3f);
     c >>= 6;
     }
    *pt = PRIV(utf8_table2)[j] | c;
    p8++;
    }
  *output_used = used;
  }

/* All done. */

EXIT:
if (inbuffer != NULL && inbuffer != stack_inbuffer)
  context->free(inbuffer, context->memory_data);

if (outbuffer != NULL && outbuffer != stack_outbuffer &&
    outbuffer != output_string)
  context->free(outbuffer, context->memory_data);

return yield;

/* Common overflow code */

OVERFLOW:
yield = B2PF_ERROR_OVERFLOW;
goto EXIT;
}

/* End of b2pf.c */
